modules/core/04-channel/keeper/upgrade.go

```diff
@@ -110,6 +110,29 @@ func (k Keeper) WriteUpgradeTryChannel(
 	emitChannelUpgradeTryEvent(ctx, portID, channelID, channel, proposedUpgrade)
 }
 
+// WriteUpgradeCancelChannel writes a channel which has canceled the upgrade process. Auxiliary upgrade state is
+// also deleted.
+func (k Keeper) WriteUpgradeCancelChannel(ctx sdk.Context, portID, channelID string) error {
+	defer telemetry.IncrCounter(1, "ibc", "channel", "upgrade-cancel")
+
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		panic(fmt.Sprintf("could not find existing channel when cancelling channel upgrade, channelID: %s, portID: %s", channelID, portID))
+	}
+
+	upgrade, found := k.GetUpgrade(ctx, portID, channelID)
+	if !found {
+		panic(fmt.Sprintf("could not find existing upgrade when cancelling channel upgrade, channelID: %s, portID: %s", channelID, portID))
+	}
+
+	if err := k.restoreChannel(ctx, portID, channelID); err != nil {
+		return errorsmod.Wrap(types.ErrUpgradeRestoreFailed, fmt.Sprintf("failed to restore channel, channelID: %s, portID: %s", channelID, portID))
+	}
```

Comment:
do we want to just panic here too? I think mixing errors/panics is worse than using one or the other. Is a regular state reversion an okay outcome if a restore fails?

The current implementation of restore is using an error return, but I guess this could be a panic as well

Reply:
i like the idea of returning an error, because as we discussed before then it is clear from the method signature that `WriteUpgrade...` could fail.